cpp/src/arrow/util/hashing_benchmark.cc

```diff
@@ -111,13 +123,238 @@ static void HashLargeStrings(benchmark::State& state) {  // NOLINT non-const ref
+  int32_t total_string_size = str_vals->total_values_length();
+
+  while (state.KeepRunning()) {
+    ASSERT_OK_AND_ASSIGN(Datum hash_result,
+                         compute::CallFunction("fast_hash_64", {values_array}));
+    benchmark::DoNotOptimize(hash_result);
+  }
+
+  state.SetBytesProcessed(state.iterations() *
+                          ((values_array->length() * sizeof(int64_t)) +
+                           (total_string_size) +
+                           (values_array->length() * sizeof(int64_t))));
+  state.SetItemsProcessed(state.iterations() * 3 * values_array->length());
+}
+
+static void FastHash64StructLargeStrings(
+    benchmark::State& state) {  // NOLINT non-const reference
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<StructArray> values_array,
+                       MakeStructArray(10000, 120, 2000));
+
+  // 2nd column (index 1) is a string column, which has offset type of int32_t
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> values_second,
+                       values_array->GetFlattenedField(1));
+  std::shared_ptr<StringArray> str_vals =
+      std::static_pointer_cast<StringArray>(values_second);
+  int32_t total_string_size = str_vals->total_values_length();
+
+  while (state.KeepRunning()) {
+    ASSERT_OK_AND_ASSIGN(Datum hash_result,
+                         compute::CallFunction("fast_hash_64", {values_array}));
+    benchmark::DoNotOptimize(hash_result);
+  }
+
+  state.SetBytesProcessed(state.iterations() *
+                          ((values_array->length() * sizeof(int64_t)) +
+                           (total_string_size) +
+                           (values_array->length() * sizeof(int64_t))));
+  state.SetItemsProcessed(state.iterations() * 3 * values_array->length());
+}
+
+static void FastHash64Map(benchmark::State& state) {  // NOLINT non-const reference
+  constexpr int64_t test_size = 10000;
+  auto test_keys = hashing_rng.String(test_size, 2, 20, /*null_probability=*/0);
+  auto test_vals = hashing_rng.Int64(test_size, 0, std::numeric_limits<int64_t>::max());
+  auto test_keyvals = hashing_rng.Map(test_keys, test_vals, test_size);
+
+  auto key_arr = std::static_pointer_cast<StringArray>(test_keys);
+  int32_t total_key_size = key_arr->total_values_length();
+  int32_t total_val_size = test_size * sizeof(int64_t);
+
+  while (state.KeepRunning()) {
+    ASSERT_OK_AND_ASSIGN(Datum hash_result,
+                         compute::CallFunction("fast_hash_64", {test_keyvals}));
+    benchmark::DoNotOptimize(hash_result);
+  }
+
+  state.SetBytesProcessed(state.iterations() * (total_key_size + total_val_size));
+  state.SetItemsProcessed(state.iterations() * 2 * test_size);
+}
+
 // ----------------------------------------------------------------------
 // Benchmark declarations
 
-BENCHMARK(HashIntegers);
+// Directly uses "Hashing" hash functions from hashing.h (xxHash)
 BENCHMARK(HashSmallStrings);
 BENCHMARK(HashMediumStrings);
 BENCHMARK(HashLargeStrings);
+BENCHMARK(HashIntegers32);
```

Comment:
So, you shouldn't benchmark compute functions outside the `arrow/compute` directory (these benchmarks must be buildable with compute disabled). Meaning, you probably need to create a separate `src/arrow/compute/kernels/scalar_hash_benchmark.cc`.